### ISSUE TITLE
SDV gets stuck during model fit (3 level deep foreign key dependency)

### ISSUE THREAD

**Issue by swapniltw** [sdv-dev/SDV#1085]
### Environment Details

Please indicate the following details about the environment in which you found the bug:

* SDV version: 0.17.0
* Python version: 3.8.13
* Operating System: MacOS Monterey Version 12.3

### Error Description

First of all, sincerely thank you for the great work on SDV.

Here's our use case. We are trying to train a set of relational tables(under schema 'customer2') as seen in the attached ER data-model at the end. We have simulated a 3 level deep foreign key dependency.

When we call `model.fit()`, we get the following performance warning trace. It then just stays stuck and never completes. Our data is very small (<50 records per table). Could you please let us know what might be causing this issue, and also direct us to some potential solutions?

### Steps to reproduce

The corresponding metadata is attached at the end.  If required, we can provide the sql file to create the schema, tables as well as mock data.

```
model = SDV()
model.fit(metadata, tables) # tables is a dictionary of dataframes
```

And it gets stuck after the following warning trace

```
my_virtual_env/lib/python3.8/site-packages/rdt/transformers/categorical.py:187: UserWarning: The data contains 10 new categories that were not seen in the original data (examples: {'IL', 'MS'}). Assigning them random values. If you want to model new categories, please fit the transformer again with the new data.
  warnings.warn(
my_virtual_env/lib/python3.8/site-packages/rdt/transformers/categorical.py:187: UserWarning: The data contains 10 new categories that were not seen in the original data (examples: {'Mankato', 'Chicago'}). Assigning them random values. If you want to model new categories, please fit the transformer again with the new data.
  warnings.warn(
my_virtual_env/lib/python3.8/site-packages/rdt/transformers/categorical.py:187: UserWarning: The data contains 10 new categories that were not seen in the original data (examples: {'United States'}). Assigning them random values. If you want to model new categories, please fit the transformer again with the new data.
  warnings.warn(
my_virtual_env/lib/python3.8/site-packages/sdv/relational/hma.py:255: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  table.reset_index(inplace=True)
my_virtual_env/lib/python3.8/site-packages/sdv/relational/hma.py:258: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  table[name] = values

```

Another thing worth mentioning is that this issue started arising only once we introduced the 3rd level foreign key chaining(profile_address->address->address_type->address_status). The model.fit() worked just fine when we had 2nd level foreign key chaining(profile_address->address->address_type).

For what it's worth, I'm working on [MacBook(Apple M1 Pro) with 32GB RAM.

Do let me know if you need any more details and I'd be happy to add. Thanks!

---------------------------------------------

[sample_metadata.txt](https://github.com/sdv-dev/SDV/files/9864151/sample_metadata.txt)

<img width="857" alt="sample_er_diagram" src="https://user-images.githubusercontent.com/109548435/197885636-7af6272e-afdf-4865-8fef-798057f4ff39.png">

**Comment by npatki** [comment]
Hi @swapniltw nice to meet you! The default `SDV()` command uses our `HMA1` model. You may find it useful to go through the blue box in the [HMA1 User Guide](https://sdv.dev/SDV/user_guides/relational/hma1.html) for more information about which datasets are suitable for this model.

As a potential workaround for your particular schema, you may have some success after modifying the data slightly --
1. Denormalize the `address_type` and `address_status` tables into the existing `address` table. As a result, the `address` table should contain only 2 additional columns: `address_type_description` and `address_status_description`
2. Model the 3 tables: `profile_address`, `profile` and (denormalized) `address` table
3. Later after synthesizing new data, you can normalize the `address` table again into `address_type` and `address_status`

For other solutions, feel free to contact us at info@sdv.dev. You can also follow along #830 that we're tracking for general improvements to modeling & sampling.

**Comment by swapniltw** [comment]
Thank you @npatki for the denormalization recommendation. Will reach out if there are any further questions.

**Comment by moidan4** [comment]
I'm facing the same issue, this how my metadata looks like: 
The process is loading non-stop
. 

```
DEM_METADATA = {
    'tables': {
        'customer': {
            'primary_key': 'customer_id',
            'fields': {
                'customer_id': {
                    'type': 'id',
                    'subtype': 'integer'
                },
                'ssn': {
                    'type': "categorical",
                    # 'pii': true,
                    'pii_category': "ssn"
                },
                'first_name': {
                    'type': "categorical",
                    # 'pii': true,
                    'pii_category': "first_name"
                },
                'last_name': {
                    'type': "categorical",
                    # 'pii': true,
                    'pii_category': "last_name"
                }
            }
        },

        'contract': {
            'primary_key': 'contract_id',
            'fields': {
                'contract_id': {
                    'type': 'id',
                    'subtype': 'integer'
                },
                'customer_id': {
                    'type': 'id',
                    'subtype': 'integer',
                    'ref': {
                    'field': 'customer_id',
                    'table': 'customer'
                   }
                },
                'contract_ref_id': {
                    'type': 'numerical',
                    # 'pii': true,
                    'subtype': 'integer',
                },
                'associated_line': {
                    'type': 'categorical',
                    # 'pii': true,
                    'pii_category': 'phone_number'
                },
                'contract_description': {
                    'type': 'categorical',
                    
                },
                'from_date': {
                    'type': 'datetime',
                    'format': '%d/%m/%Y %H:%M:%S'
                },
                'to_date': {
                    'type': 'datetime',
                    'format': '%d/%m/%Y %H:%M:%S'
                },
                'associated_line_fmt': {
                    'type': 'numerical',
                    'subtype': 'integer'
                    
                }
            }
        },
     
        'subscriber': {
            'primary_key': 'subscriber_id',
            'fields': {
                'subscriber_id': {
                    'type': 'id',
                    'subtype': 'integer',
                    'ref': {
                    'field': 'contract_id',
                    'table': 'contract'
                   }
                },
                'msisdn': {
                    'type': 'id',
                    'subtype': 'integer',
                    
                },
                'imsi': {
                    'type': 'numerical',
                   
                    'subtype': 'integer'
                },
                'sim': {
                    'type': 'numerical',
                   
                    'subtype': 'integer'
                },
                'first_name': {
                    'type': "categorical",
                    # 'pii': true,
                    'pii_category': "first_name"
                },
                'last_name': {
                    'type': "categorical",
                    # 'pii': true,
                    'pii_category': "last_name"
                },
                'subscriber_type': {
                    'type': 'numerical',
                   
                    'subtype': 'integer'
                },
                'vip_status': {
                    'type': 'categorical',
                   
                }
            }
        },
      
    }
}
```

**Comment by npatki** [comment]
Hi @moidan4, judging by the smaller # of columns and depth of your metadata, I don't know if the processing time is due to the same issue as the original commenter. 

Would you be able to file a new issue to look into this if you're still having issues?

**Comment by npatki** [comment]
Hello, seems like the conversation in this issue has died down, so I'll mark it as closed. 

If you are still having issues with this, please feel free to reply and I can reopen for investigation.
